clang/test/SemaTemplate/concepts-out-of-line-def.cpp

```diff
@@ -599,3 +599,39 @@ template <class DerT>
 unsigned long DerivedCollection<DerTs...>::index() {}
 
 } // namespace GH72557
+
+namespace GH102320 {
+
+template <class, class>
+concept Constrained = true;
+
+template <class T> class C {
+  template <Constrained<T>> class D;
+  template <class U>
+    requires Constrained<T, U>
+  class E;
+};
+
+template <class T> template <Constrained<T>> class C<T>::D {};
+
+template <class T>
+template <class U>
+  requires Constrained<T, U>
+class C<T>::E {};
+
+#if 0
+// FIXME: Is it conforming? Only Clang rejects it in every released version.
+template <>
+template <Constrained<int> T>
+class C<int>::D<T> {};
+#endif
+
```

Comment:
I don't think we need this regression. As pointed in this [comment](https://github.com/llvm/llvm-project/pull/102554#discussion_r1710667937), we should not transform `ConceptSpecializationExpr` which makes the depth incorrect. I tested locally and my origin approach also makes sense to this case.

Reply:
I DIDN’T say this is a regression, but rather we’ve been rejecting it in many versions.

The function per se is called SubstituteConstraintExpressionWithoutSatisfaction, so not transforming the constraint appears to go against the intention of it to me.

Reply:
I think if we really need to end up with an approach that by not transforming it, we should do the check/special-casing inside getTemplateInstantiationArgs(), rather than doing that at the call site.

Reply:
Actually there seems to be a slight discrepancy among the four compilers here:
https://gcc.godbolt.org/z/nxPoeGE5b
w/o constraints, MSVC and Clang reject such a pattern, while GCC and EDG accept it.
w/ constraints, only Clang complains. So I'm unclear about what is supposed to happen here.
@cor3ntin do you have any ideas?

Reply:
@zyn0217 I need more time to think about the concept case, but the first case looks definitively like a bug. The following looks related https://github.com/llvm/llvm-project/issues/53139